### PR TITLE
Strip React.PropTypes type checking code in production

### PR DESCRIPTION
**Pull request by gaearon**
This is #6401 revived.
It’s a major change that can go in 16 now that we warn for calls in 15.

This gives us -1K from isomorphic React post gzip.

```
   raw     gz Compared to last run                                             
     =      = build/react-dom-fiber.js                                         
     =      = build/react-dom-fiber.min.js                                     
     =      = build/react-dom-server.js                                        
     =      = build/react-dom-server.min.js                                    
     =      = build/react-dom.js                                               
     =      = build/react-dom.min.js                                           
 +1200   +213 build/react-with-addons.js                                       
 -3703   -999 build/react-with-addons.min.js                                   
 +1200   +216 build/react.js                                                   
 -3703   -988 build/react.min.js   
```

### CODE REVIEW

**Comment by gaearon** [comment]
Also pushed a revert for warning added in #7132.

File: src/isomorphic/classic/types/ReactPropTypes.js [hunk 1]
```diff
@@ -68,25 +68,57 @@ var warning = require('warning');
...
+} else {
+  var productionTypeChecker = function() {
+    invariant(
+      false,
+      'React.PropTypes type checking code is stripped in production.'
+    );
+  };
+  productionTypeChecker.isRequired = productionTypeChecker;
+  var getProductionTypeChecker = () => productionTypeChecker;
+  // Keep in sync with production version above
```
Comment (fson) [inline src/isomorphic/classic/types/ReactPropTypes.js]: I believe this should say "Keep in sync with the development version above."

Comment (gaearon) [inline src/isomorphic/classic/types/ReactPropTypes.js]: Thanks

**Comment by gaearon** [comment]
The PR is fine, don’t be fooled by 0.007% decrease in coverage..

**Comment by zpao** [comment]
FWIW, I just changed the threshold for coveralls to -1% so hopefully we don't get too much noise from failures there.

File: src/isomorphic/classic/types/ReactPropTypes.js [hunk 2]
```diff
@@ -68,25 +68,57 @@ var warning = require('warning');
...
+    invariant(
+      false,
+      'React.PropTypes type checking code is stripped in production.'
+    );
+  };
+  productionTypeChecker.isRequired = productionTypeChecker;
+  var getProductionTypeChecker = () => productionTypeChecker;
+  // Keep in sync with development version above
+  var ReactPropTypes = {
+    array: getProductionTypeChecker(),
```
Comment (vjeux) [inline src/isomorphic/classic/types/ReactPropTypes.js]: What's the point of calling it here? Why not just put `productionTypeChecker
`?

File: src/isomorphic/classic/types/ReactPropTypes.js [hunk 3]
```diff
@@ -120,41 +152,16 @@ function PropTypeError(message) {
...
-  }
   function checkType(
     isRequired,
     props,
     propName,
     componentName,
     location,
-    propFullName,
-    secret
+    propFullName
```
Comment (vjeux) [inline src/isomorphic/classic/types/ReactPropTypes.js]: you can leave the trailing comma :)

File: src/isomorphic/classic/types/__tests__/ReactPropTypesProduction-test.js [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    var props = {testProp: value};
+    expect(() => {
+      declaration(
+        props,
+        'testProp',
+        'testComponent',
+        ReactPropTypeLocations.prop
+      );
+    }).toThrow(
+      /React.PropTypes type checking code is stripped in production./
```
Comment (vjeux) [inline src/isomorphic/classic/types/__tests__/ReactPropTypesProduction-test.js]: why use a regex here and not just a string? also if it's a regex you'd probably want to escape the dot

File: src/isomorphic/classic/types/ReactPropTypes.js (hunk 2 above)
Comment (gaearon) [inline src/isomorphic/classic/types/ReactPropTypes.js]: I'm trying to mirror the structure above exactly so it's easy to keep them in sync.

File: src/isomorphic/classic/types/__tests__/ReactPropTypesProduction-test.js [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    });
+  });
+
+  describe('Custom validator', function() {
+    beforeEach(function() {
+      jest.resetModuleRegistry();
+    });
+
+    it('should not have been called', function() {
+      var spy = jasmine.createSpy();
```
Comment (vjeux) [inline src/isomorphic/classic/types/__tests__/ReactPropTypesProduction-test.js]: why not jest.fn?

**Comment by gaearon** [comment]
Closing, will resubmit as another PR.

**Comment by gaearon** [comment]
#8066

**Pull request closed without merging**